src/thermo/Phase.cpp

```diff
@@ -256,42 +256,125 @@ std::string Phase::speciesSPName(int k) const
     return m_name + ":" + speciesName(k);
 }
 
+std::map<std::string, size_t> Phase::nativeState() const
+{
+    if (isPure()) {
+        if (isCompressible()) {
+            return { {"T", 0}, {"density", 1} };
+        } else {
+            return { {"T", 0}, {"P", 1} };
+        }
+    } else {
+        if (isCompressible()) {
+            return { {"T", 0}, {"density", 1}, {"Y", 2} };
+        } else {
+            return { {"T", 0}, {"P", 1}, {"Y", 2} };
+        }
+    }
+}
+
+vector<std::string> Phase::fullStates() const
+{
+    if (isPure()) {
+        if (isCompressible()) {
+            return {"TD", "TP", "UV", "DP", "HP", "SP", "SV"};
+        } else {
+            return {"TP", "HP", "SP"};
+        }
+    } else {
+        if (isCompressible()) {
+            return {"TDX", "TDY", "TPX", "TPY", "UVX", "UVY", "DPX", "DPY",
+                    "HPX", "HPY", "SPX", "SPY", "SVX", "SVY"};
+        } else {
+            return {"TPX", "TPY", "HPX", "HPY", "SPX", "SPY"};
+        }
+    }
+}
+
+vector<std::string> Phase::partialStates() const
+{
+    if (isPure()) {
+        return {};
+    } else {
+        if (!isCompressible()) {
+            return {"TP", "HP", "SP"};
+        } else {
+            return {"TD", "TP", "UV", "DP", "HP", "SP", "SV"};
+        }
+    }
+}
+
+size_t Phase::nStates() const {
+    if (isPure()) {
+        return 2;
+    } else {
+        return nSpecies() + 2;
+    }
+}
+
 void Phase::saveState(vector_fp& state) const
 {
-    state.resize(nSpecies() + 2);
+    state.resize(nStates());
     saveState(state.size(), &state[0]);
 }
+
 void Phase::saveState(size_t lenstate, doublereal* state) const
 {
-    state[0] = temperature();
-    state[1] = density();
-    getMassFractions(state + 2);
+    auto native = nativeState();
+
+    // function assumes default definition of nativeState
+    state[native["T"]] = temperature();
+    if (isCompressible()) {
+        state[native["density"]] = density();
+    } else {
+        state[native["P"]] = pressure();
+    }
+    if (!isPure()) {
+        if (native.find("Y") == native.end()) {
+            getMoleFractions(state + native["X"]);
+        } else {
+            getMassFractions(state + native["Y"]);
+        }
+    }
```

Comment:
To avoid accidentally reading 0 for non-existent keys:
```c++
        if (native.count("X")) {
            getMoleFractions(state + native["X"]);
        } else if (native.count("Y") {
            getMassFractions(state + native["Y"]);
        }
```
And actually, this structure negates the need to check on `isPure()`.

Reply:
Nice.